src/lib.rs

```diff
@@ -171,15 +171,15 @@ impl<T> CacheKey for Key<T> where T: 'static {
 }
 
 /// Resource store. Responsible for holding and presenting resources.
-pub struct Store {
+pub struct Store<C> {
```

Comment:
Having a `Store` bound to a context is not something cool, since libraries relying on a `Store` might have to hack around or even expose a type, disabling end-users to provide a context type for their own use.

Reply:
Yeah, I don't like it either, really.  It works for me only because I control the whole stack.

The difficulty with getting rid of it is this field in `Store<C>`:

```rust
  // contains all metadata on resources (reload functions, last time updated, etc.)
  metadata: HashMap<PathBuf, ResMetaData<C>>,
```

Which contains this:

```rust
/// Meta data about a resource.
struct ResMetaData<C> {
  /// Function to call each time the resource must be reloaded.
  on_reload: Box<Fn(&mut Store<C>, &mut C) -> Result<(), Box<Error>>>,
  /// The last time the resource was updated.
  last_update_instant: Instant,
}
```

We would need to make some way of making it generic on C without affecting the `Store`... like you say, it might be able to be an associated type, but that might need defining a new trait.

Reply:
A solution would be to embed the the context in a `Rc<RefCell<_>>`, but that won’t work if you cannot clone your context.

I think your solution might work if libraries use polymorphic objects, like:

    fn foo<T>(store: &mut Store<T>) where Context: AsRef<T>;